igvc_platform/src/motor_controller/main.cpp

```diff
@@ -1,247 +1,286 @@
+double p_l, p_r, d_l, d_r, i_l, i_r; // PID Values
 
+/**
+get current motor command from the /motors topic
+
+@param[in] msg current message on the /motors topic
+*/
 void cmdCallback(const igvc_msgs::velocity_pair::ConstPtr& msg)
 {
   current_motor_command = *msg;
 }
 
-std::string toBoundedString(double input)
+/**
+Sets PID values
+*/
+void setPID(EthernetSocket& sock, ros::Rate &rate)
 {
-  std::stringstream stream;
-  stream << std::fixed << std::setprecision(precision) << input;
-  return stream.str();
-}
+  bool valid_values = false; // pid values have been set correctly
 
-bool validateValues(std::string ret, int loc, int end, double left, double right)
-{
-  std::vector<std::string> vals = split(ret.substr(loc + 1, end), ',');
-  ROS_INFO_STREAM("Successfully set p values");
-  bool valid_values = true;
-  if (vals.size() == 2)
+  /* This is the buffer where we will store the request message. */
+  uint8_t requestbuffer[256];
+  size_t message_length;
+  bool status;
+
+  /* allocate space for the request message to the server */
+  RequestMessage request = RequestMessage_init_zero;
+
+  /* Create a stream that will write to our buffer. */
+  pb_ostream_t ostream = pb_ostream_from_buffer(requestbuffer, sizeof(requestbuffer));
+
+  /* indicate that pid fields will contain values */
+  request.has_p_l = true;
+  request.has_p_r = true;
+  request.has_i_l = true;
+  request.has_i_r = true;
+  request.has_d_l = true;
+  request.has_d_r = true;
+
+  /* fill in the message fields */
+  request.p_l = static_cast<float>(p_l);
+  request.p_r = static_cast<float>(p_r);
+  request.i_l = static_cast<float>(i_l);
+  request.i_r = static_cast<float>(i_r);
+  request.d_l = static_cast<float>(d_l);
+  request.d_r = static_cast<float>(d_r);
+
+  /* encode the protobuffer */
+  status = pb_encode(&ostream, RequestMessage_fields, &request);
+  message_length = ostream.bytes_written;
+
+  /* check for any errors.. */
+  if (!status)
   {
-    valid_values = (stod(vals.at(0)) == left);
-    valid_values = (stod(vals.at(1)) == right) && valid_values;
+      ROS_ERROR_STREAM("Encoding failed: " << PB_GET_ERROR(&ostream));
+      ros::shutdown();
+  }
+
+  size_t n; // n is the response from socket: 0 means connection closed, otherwise n = num bytes read
+  uint8_t buffer[256]; // buffer to read response into
+  // unsigned char responsebuffer[256];
+
+  /* Send PID values via ethernet and recieve response to ensure proper setting */
+  while (ros::ok() && !valid_values)
+  {
+    sock.sendMessage(reinterpret_cast<char*>(requestbuffer), message_length);
+
+    memset(buffer, 0, sizeof(buffer));
+    n = sock.readMessage(buffer); // blocks until data is read
+    // memcpy(responsebuffer, buffer, sizeof(responsebuffer));
+
+    if (n == 0)
+    {
+        ROS_ERROR_STREAM("Connection closed by server");
+        ros::shutdown();
+    }
+
+    /* Allocate space for the decoded message. */
+    ResponseMessage response = ResponseMessage_init_zero;
```

Comment:
Doesn't the default constructor take care of this?

Reply:
I don't think a default constructor is defined for `ResponseMessage`. Also, I think it's better to make it explicit that all fields will be initialized to zero.